# Rapid and parallel adaptive mutations in spike S1 drive clade success in SARS-CoV-2
**Kathryn Kistler** <sup>1,2</sup>, **John Huddleston** <sup>2</sup> , **Trevor Bedford** <sup>1,2,3</sup> <br />
<sup>1</sup> *Molecular and Cellular Biology Program, University of Washington, Seattle, United States* <br />
<sup>2</sup> *Vaccine and Infectious Disease Division, Fred Hutchinson Cancer Research Center, Seattle, United States*
<sup>3</sup> *Howard Hughes Medical Institute, Seattle, WA, United States*

## Abstract
Given the importance of variant SARS-CoV-2 viruses with altered receptor-binding or antigenic phenotypes, we sought to quantify the degree to which adaptive evolution is driving accumulation of mutations in the SARS-CoV-2 genome.
Here we assessed adaptive evolution across genes in the SARS-CoV-2 genome by correlating clade growth with mutation accumulation as well as by comparing rates of nonsynonymous to synonymous divergence, clustering of mutations across the SARS-CoV-2 phylogeny and degree of convergent evolution of individual mutations.
We find that spike S1 is the focus of adaptive evolution, but also identify positively-selected mutations in other genes that are sculpting the evolutionary trajectory of SARS-CoV-2.
Adaptive changes in S1 accumulated rapidly, resulting in a remarkably high ratio of nonsynonymous to synonymous divergence that is 2.5X greater than that observed in HA1 at the beginning of the 2009 H1N1 pandemic.


[![DOI](https://zenodo.org/badge/388629418.svg)](https://zenodo.org/badge/latestdoi/388629418)
