# Rapid and parallel adaptive mutations in spike S1 drive clade success in SARS-CoV-2
**Kathryn Kistler** <sup>1,2</sup>, **John Huddleston** <sup>2</sup> , **Trevor Bedford** <sup>1,2</sup> <br />
<sup>1</sup> *Molecular and Cellular Biology Program, University of Washington, Seattle, United States* <br />
<sup>2</sup> *Vaccine and Infectious Disease Division, Fred Hutchinson Cancer Research Center, Seattle, United States*

## Abstract
Despite the appearance of variant SARS-CoV-2 viruses with altered receptor-binding or antigenic phenotypes, traditional methods for detecting adaptive evolution from sequence data do not pick up strong signals of positive selection.
Here, we present a new method for identifying adaptive evolution on short evolutionary time scales with densely-sampled populations.
We apply this method to SARS-CoV-2 to perform a comprehensive analysis of adaptively-evolving regions of the genome.
We find that spike S1 is a focal point of adaptive evolution, but also identify positively-selected mutations in other genes that are sculpting the evolutionary trajectory of SARS-CoV-2.
The adaptive changes in S1 accumulated rapidly and in parallel in several lineages, resulting in a remarkably high ratio of nonsynonymous to synonymous divergence that is 2.5 times greater than that observed in HA1 at the beginning of the 2009 H1N1 pandemic.
