{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 351,
   "metadata": {},
   "outputs": [],
   "source": [
    "from augur.utils import json_to_tree\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from collections import Counter\n",
    "from sklearn import linear_model\n",
    "from sklearn.model_selection import ShuffleSplit\n",
    "from sklearn.metrics import mean_squared_error\n",
    "import requests\n",
    "import re\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get global ncov tree and convert to Bio Phylo format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 352,
   "metadata": {},
   "outputs": [],
   "source": [
    "tree_url = \"https://data.nextstrain.org/ncov_global.json\"\n",
    "tree_json = requests.get(tree_url).json()\n",
    "tree = json_to_tree(tree_json)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Download entropy manually from nextstrain.org/ncov/global. Find the 0.5% of sites with highest entropy throughout genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 312,
   "metadata": {},
   "outputs": [],
   "source": [
    "entropy_file = 'nextstrain_ncov_global_diversity.tsv'\n",
    "entropy_df = pd.read_csv(entropy_file, sep='\\t')\n",
    "# genome is 29902 nt long, \n",
    "# which would be roughly 9967 codons (but this isn't quite right since there are non-coding regions)\n",
    "# estimate about 9950 codons -> top 1% would 99.5 codons... so take 100 highest entropy values\n",
    "top_entropy_df = entropy_df.nlargest(100,'entropy')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read in sequence file, that will be used to find the genotype of all members of a clade\n",
    "\n",
    "aws s3 cp s3://nextstrain-ncov-private/global_subsampled_sequences.fasta.xz ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 353,
   "metadata": {},
   "outputs": [],
   "source": [
    "sequences_file = 'global_subsampled_sequences.fasta'\n",
    "record_dict = SeqIO.to_dict(SeqIO.parse(sequences_file, \"fasta\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make dataframe where each row is a clade and columns contain information about logistic growth rate, how many S1 mutations from root to clade, and percentage of tips in clade that are genotype X\n",
    "\n",
    "First, need to find genotype of isolates at desired positions (of highest entropy). Do this for all tips and store in a dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in reference file and find genome position for each codon in the top entropy sites\n",
    "\n",
    "# find the start position of the codon encoding each entropic site\n",
    "genome_location_of_entropic_sites = {}\n",
    "\n",
    "for k,v in top_entropy_df.iterrows():\n",
    "\n",
    "    for record in SeqIO.parse(open(\"reference_seq_edited.gb\",\"r\"), \"genbank\"):\n",
    "        for feature in record.features:\n",
    "            if feature.type == 'CDS':\n",
    "                if feature.qualifiers['gene'][0] == v['gene']:\n",
    "                    mut_location_start = feature.location.start + int(v['position'] -1)*3\n",
    "#                     print((record.seq[mut_location_start:mut_location_start+3]).translate())\n",
    "                    genome_location_of_entropic_sites[mut_location_start] = f\"{v['gene']}_{v['position']}\"\n",
    "\n",
    "                    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "metadata": {},
   "outputs": [],
   "source": [
    "# initialize dictionary to store genotype info\n",
    "tip_genotypes = {}\n",
    "\n",
    "for k,v in record_dict.items():\n",
    "    this_tip = {}\n",
    "    #store genotype at each entropic site\n",
    "    for nt_start, gene_codon in genome_location_of_entropic_sites.items():\n",
    "        nt_codon = v.seq[nt_start:nt_start+3]\n",
    "        if all([characters in ['A', 'C', 'G', 'T'] for characters in nt_codon]):\n",
    "            this_tip[gene_codon] = str(nt_codon.translate())\n",
    "        elif all([characters in ['-'] for characters in nt_codon]):\n",
    "            this_tip[gene_codon] = '-'\n",
    "        else:\n",
    "            this_tip[gene_codon] = None\n",
    "\n",
    "    tip_genotypes[k] = this_tip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 356,
   "metadata": {},
   "outputs": [],
   "source": [
    "#initiate list to store all clade info\n",
    "clade_stats = []\n",
    "\n",
    "#Only want to look at clades, don't care about tips\n",
    "for node in tree.find_clades(terminal=False):\n",
    "    \n",
    "    # only care about clade if logistic growth value is associated\n",
    "    # find stored logistic growth value\n",
    "    if \"logistic_growth\" in node.node_attrs:\n",
    "        logistic_growth = node.node_attrs[\"logistic_growth\"][\"value\"]\n",
    "        \n",
    "        #all S1 muts from root to clade already stored as json value\n",
    "        if \"S1_mutations\" in node.node_attrs:\n",
    "            s1_mutations = node.node_attrs[\"S1_mutations\"][\"value\"]\n",
    "        else:\n",
    "            s1_mutations = None\n",
    "\n",
    "        # find all tips in this clade\n",
    "        tips_in_clade = node.get_terminals()\n",
    "        tip_names_in_clade = [tip.name for tip in tips_in_clade]\n",
    "        \n",
    "        # tally the genotype of each tip within the clade\n",
    "        clade_genotype_tally = {}\n",
    "        for tip in tip_names_in_clade:\n",
    "            for site, aa in tip_genotypes[tip].items():\n",
    "                # ignore the isolates that were not sequenced\n",
    "                if aa != None:\n",
    "                    if site in clade_genotype_tally.keys():\n",
    "                        clade_genotype_tally[site]+=[aa]\n",
    "                    else:\n",
    "                        clade_genotype_tally[site] = [aa]\n",
    "        \n",
    "        # find proportion of tips that have given genotype\n",
    "        clade_genotype_freqs = {}\n",
    "        for k, v in clade_genotype_tally.items():\n",
    "            counts = Counter(v)\n",
    "            for item, count in counts.items():\n",
    "                site_genotype = k+item\n",
    "                clade_genotype_freqs[site_genotype] = count/len(v)\n",
    "\n",
    "        \n",
    "        clade_stats.append({'clade': node.name, 'logistic_growth': logistic_growth, \n",
    "                            'num_s1_muts': s1_mutations, **clade_genotype_freqs})\n",
    "\n",
    "\n",
    "clade_stats_df = pd.DataFrame(clade_stats).fillna(0)\n",
    "                \n",
    "\n",
    "    \n",
    "\n",
    "    \n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter to only mutations that occur at least 4 times on the tree. And don't allow reversion mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 357,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/katekistler/anaconda3/envs/ncov/lib/python3.6/site-packages/Bio/Seq.py:2859: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  BiopythonWarning,\n"
     ]
    }
   ],
   "source": [
    "# find root sequence, in order to exclude reversions\n",
    "\n",
    "reference_sequence = {}\n",
    "\n",
    "for record in SeqIO.parse(open(\"reference_seq_edited.gb\",\"r\"), \"genbank\"):\n",
    "    for feature in record.features:\n",
    "        if feature.type == 'CDS':\n",
    "            gene_seq = feature.location.extract(record.seq).translate()\n",
    "            reference_sequence[feature.qualifiers['gene'][0]] = gene_seq\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 358,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_mutations = []\n",
    "\n",
    "all_positions = []\n",
    "\n",
    "\n",
    "#only look at mutations on internal branches\n",
    "for node in tree.find_clades(terminal=False):\n",
    "    \n",
    "    num_descendents = len(node.get_terminals())\n",
    "\n",
    "    # mutations must be propagated to at least a small clade\n",
    "    if num_descendents >=10:\n",
    "        # look just at nonsyn muts\n",
    "        if hasattr(node, 'branch_attrs'):\n",
    "            for gene, mut_list in node.branch_attrs[\"mutations\"].items():\n",
    "                if gene!= 'nuc':\n",
    "                    for mut in mut_list:\n",
    "                        gene_mutation = f'{gene}_{mut[1:]}'\n",
    "                        gene_position = f'{gene}_{mut[1:-1]}'\n",
    "                        all_mutations.append(gene_mutation)\n",
    "                        all_positions.append(gene_position)\n",
    "\n",
    "\n",
    "\n",
    "independent_occurences_all_mutations = Counter(all_mutations)\n",
    "independent_occurences_all_positions = Counter(all_positions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 337,
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter the dataframe to only include mutations that are not reversions, and occur at least 4 times over tree\n",
    "clade_stats_filtered_df = clade_stats_df.copy()\n",
    "\n",
    "to_drop = []\n",
    "for m, count in independent_occurences_all_mutations.items():\n",
    "\n",
    "    \n",
    "    if m in clade_stats_filtered_df.columns:\n",
    "        # drop mutations that appear less than 4 independent times\n",
    "        if count<4:\n",
    "            to_drop.append(m)\n",
    "            \n",
    "for m in clade_stats_filtered_df.columns[3:]:\n",
    "    # find whether mutation is a reversion, and drop it if so\n",
    "    gene = m.split('_')[0]\n",
    "    mut = m.split('_')[1]\n",
    "\n",
    "    if reference_sequence[gene][int(mut[:-1])-1] == mut[-1]:\n",
    "        to_drop.append(m)\n",
    "\n",
    "\n",
    "clade_stats_filtered_df.drop(to_drop, inplace=True, axis=1)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Want to find muts that drive/predict logisitic growth. Intuitively, the prevalence of these muts should be enriched in clades with high logistic growth compared to clades with low logistic growth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 320,
   "metadata": {},
   "outputs": [],
   "source": [
    "def log_growth_enrichment(threshold, df):\n",
    "    # get rid of first 3 columns (clade, logistic_growth, num_s1_muts)\n",
    "    all_mut_keys = list(df.keys())[3:]\n",
    "\n",
    "    avg_prevalence_high_growth = df[df['logistic_growth']>=threshold][all_mut_keys].mean()\n",
    "    avg_prevalence_low_growth = df[df['logistic_growth']<threshold][all_mut_keys].mean()\n",
    "    \n",
    "    avg_prevalence_overall = df[all_mut_keys].mean()\n",
    "    \n",
    "\n",
    "    print(avg_prevalence_high_growth.subtract(avg_prevalence_low_growth).nlargest(20))\n",
    "    \n",
    "#     print((avg_prevalence_high_growth/avg_prevalence_overall).nlargest(20))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ORF8_73C       0.690065\n",
      "ORF1a_1708D    0.687376\n",
      "S_1118H        0.687325\n",
      "S_570D         0.686958\n",
      "S_982A         0.686948\n",
      "ORF1a_2230T    0.686314\n",
      "N_235F         0.685546\n",
      "ORF8_27*       0.684941\n",
      "ORF8_52I       0.684897\n",
      "ORF1a_1001I    0.684401\n",
      "S_501Y         0.678876\n",
      "N_3L           0.677292\n",
      "S_716I         0.669107\n",
      "S_144-         0.660464\n",
      "S_69-          0.636046\n",
      "S_681H         0.615012\n",
      "ORF1a_3675-    0.611973\n",
      "ORF1a_3676-    0.606666\n",
      "ORF1a_3677-    0.604868\n",
      "N_203K         0.554096\n",
      "dtype: float64\n"
     ]
    }
   ],
   "source": [
    "log_growth_enrichment(3.0, clade_stats_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After 501Y and nsp6 deletion, many of the hits above are just in 501Y.v1 clade. Because that clade is so large it skews the average. Look at the filtered df, which requires that the mutation has occurred at least 4 on the phylogeny, spreading to at least 10 descendents each time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S_501Y         0.314463\n",
      "ORF1a_3675-    0.290252\n",
      "ORF1a_3676-    0.290138\n",
      "ORF1a_3677-    0.289553\n",
      "S_681H         0.276434\n",
      "S_484K         0.042783\n",
      "M_82T          0.041230\n",
      "S_452R         0.037842\n",
      "S_18F          0.028800\n",
      "S_95I          0.027188\n",
      "ORF8_68-       0.010788\n",
      "ORF8_24-       0.010748\n",
      "ORF8_27-       0.010748\n",
      "ORF8_52-       0.010748\n",
      "ORF8_84-       0.010710\n",
      "ORF8_92-       0.010710\n",
      "ORF8_73-       0.010710\n",
      "S_138H         0.006039\n",
      "S_144F         0.004523\n",
      "S_477I         0.004105\n",
      "dtype: float64\n"
     ]
    }
   ],
   "source": [
    "log_growth_enrichment(3.0, clade_stats_filtered_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 339,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S_501Y         0.459344\n",
      "S_681H         0.437056\n",
      "ORF1a_3677-    0.414989\n",
      "ORF1a_3675-    0.414634\n",
      "ORF1a_3676-    0.414170\n",
      "M_82T          0.007173\n",
      "ORF8_68-       0.006133\n",
      "ORF8_52-       0.006110\n",
      "ORF8_27-       0.006110\n",
      "ORF8_24-       0.006110\n",
      "ORF8_73-       0.006089\n",
      "ORF8_92-       0.006089\n",
      "ORF8_84-       0.006089\n",
      "S_138H         0.005783\n",
      "ORF1a_3676S    0.005642\n",
      "ORF1a_3677L    0.004800\n",
      "ORF1a_3675T    0.004634\n",
      "S_95I          0.004116\n",
      "N_3Q           0.003649\n",
      "ORF9b_32H      0.002812\n",
      "dtype: float64\n"
     ]
    }
   ],
   "source": [
    "log_growth_enrichment(1.0, clade_stats_filtered_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Want to fit a multiple linear regression to see which mutations best predict logisitic growth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 340,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ORF8_92- 2386889.7506320127\n",
      "ORF8_27- 61597.79208677484\n",
      "ORF8_24- 34653.172970179105\n",
      "ORF8_52- 27754.962856982314\n",
      "N_204- 25547.223332664355\n",
      "N_203- 25547.223317929474\n",
      "ORF1a_1640H 6947.862432278775\n",
      "N_3G 2916.68335020463\n",
      "N_3N 1423.3744379028765\n",
      "N_202I 703.3072557435502\n",
      "ORF1b_218S 560.4135788484829\n",
      "N_203S 509.31792933764126\n",
      "S_732S 455.68478078565056\n",
      "S_716V 227.8339048337713\n",
      "M_82V 212.26524559558314\n",
      "S_215N 158.6357238632068\n",
      "S_142V 148.37199179189588\n",
      "ORF8_52K 135.73354802351815\n",
      "S_26L 107.14246619430516\n",
      "N_3E 91.13888667173268\n"
     ]
    }
   ],
   "source": [
    "# get all possible explanatory variables\n",
    "all_keys = list(clade_stats_filtered_df.keys())[2:]\n",
    "\n",
    "\n",
    "X = clade_stats_filtered_df[all_keys]\n",
    "# X = clade_stats_df[['num_s1_muts']]\n",
    "# X = clade_stats_df[['S_501Y']]\n",
    "y = clade_stats_filtered_df['logistic_growth']\n",
    "\n",
    "# make a linear regression object using sklearn\n",
    "regr = linear_model.LinearRegression()\n",
    "regr.fit(X, y)\n",
    "\n",
    "coefficients = list(regr.coef_)\n",
    "ranked_coefficients_by_index = sorted(range(len(coefficients)), key=lambda x: coefficients[x])\n",
    "top_ranked_idicies = reversed(ranked_coefficients_by_index[-20:])\n",
    "for index in top_ranked_idicies:\n",
    "    variable = all_keys[index]\n",
    "    coefficient = coefficients[index]\n",
    "    print(variable, coefficient)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 341,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S_681H 2.5001424292532115\n",
      "num_s1_muts 1.223358381816525\n",
      "S_18F 0.5535908780562372\n",
      "ORF9b_32H 0.0\n",
      "S_20A 0.0\n",
      "ORF1a_1640H 0.0\n",
      "N_2F 0.0\n",
      "S_138V 0.0\n",
      "S_138N 0.0\n",
      "S_138H 0.0\n",
      "S_138- 0.0\n",
      "ORF8_92- 0.0\n",
      "ORF9b_10H 0.0\n",
      "S_477R 0.0\n",
      "S_142V 0.0\n",
      "N_377E 0.0\n",
      "ORF1a_959T 0.0\n",
      "S_732S 0.0\n",
      "S_677L 0.0\n",
      "ORF8_84- 0.0\n"
     ]
    }
   ],
   "source": [
    "all_keys = list(clade_stats_filtered_df.keys())[2:]\n",
    "\n",
    "X = clade_stats_filtered_df[all_keys]\n",
    "y = clade_stats_filtered_df['logistic_growth']\n",
    "\n",
    "lasso = linear_model.LassoCV().fit(X, y)\n",
    "importance = np.abs(lasso.coef_)\n",
    "feature_names = np.array(all_keys)\n",
    "# plt.bar(height=importance, x=feature_names)\n",
    "\n",
    "ranked_importance_by_index = sorted(range(len(importance)), key=lambda x: importance[x])\n",
    "top_ranked_idicies = reversed(ranked_importance_by_index[-20:])\n",
    "for index in top_ranked_idicies:\n",
    "    variable = all_keys[index]\n",
    "    i = importance[index]\n",
    "    print(variable, i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Found input variables with inconsistent numbers of samples: [1566, 522]",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-116-570b41542e82>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     22\u001b[0m         \u001b[0mmodel_train\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlin_reg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX_train\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreshape\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_train\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreshape\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     23\u001b[0m         \u001b[0mpredictions_train\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmodel_train\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpredict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX_valid\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreshape\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 24\u001b[0;31m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmean_squared_error\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_train\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreshape\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpredictions_train\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     25\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/katekistler/anaconda3/envs/ncov/lib/python3.6/site-packages/sklearn/utils/validation.py\u001b[0m in \u001b[0;36minner_f\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     61\u001b[0m             \u001b[0mextra_args\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mall_args\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     62\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mextra_args\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 63\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     64\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     65\u001b[0m             \u001b[0;31m# extra_args > 0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/katekistler/anaconda3/envs/ncov/lib/python3.6/site-packages/sklearn/metrics/_regression.py\u001b[0m in \u001b[0;36mmean_squared_error\u001b[0;34m(y_true, y_pred, sample_weight, multioutput, squared)\u001b[0m\n\u001b[1;32m    334\u001b[0m     \"\"\"\n\u001b[1;32m    335\u001b[0m     y_type, y_true, y_pred, multioutput = _check_reg_targets(\n\u001b[0;32m--> 336\u001b[0;31m         y_true, y_pred, multioutput)\n\u001b[0m\u001b[1;32m    337\u001b[0m     \u001b[0mcheck_consistent_length\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_true\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_pred\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msample_weight\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    338\u001b[0m     output_errors = np.average((y_true - y_pred) ** 2, axis=0,\n",
      "\u001b[0;32m/Users/katekistler/anaconda3/envs/ncov/lib/python3.6/site-packages/sklearn/metrics/_regression.py\u001b[0m in \u001b[0;36m_check_reg_targets\u001b[0;34m(y_true, y_pred, multioutput, dtype)\u001b[0m\n\u001b[1;32m     86\u001b[0m         \u001b[0mthe\u001b[0m \u001b[0mdtype\u001b[0m \u001b[0margument\u001b[0m \u001b[0mpassed\u001b[0m \u001b[0mto\u001b[0m \u001b[0mcheck_array\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     87\u001b[0m     \"\"\"\n\u001b[0;32m---> 88\u001b[0;31m     \u001b[0mcheck_consistent_length\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_true\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_pred\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     89\u001b[0m     \u001b[0my_true\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcheck_array\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_true\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mensure_2d\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     90\u001b[0m     \u001b[0my_pred\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcheck_array\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_pred\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mensure_2d\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/katekistler/anaconda3/envs/ncov/lib/python3.6/site-packages/sklearn/utils/validation.py\u001b[0m in \u001b[0;36mcheck_consistent_length\u001b[0;34m(*arrays)\u001b[0m\n\u001b[1;32m    261\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0muniques\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    262\u001b[0m         raise ValueError(\"Found input variables with inconsistent numbers of\"\n\u001b[0;32m--> 263\u001b[0;31m                          \" samples: %r\" % [int(l) for l in lengths])\n\u001b[0m\u001b[1;32m    264\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    265\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Found input variables with inconsistent numbers of samples: [1566, 522]"
     ]
    }
   ],
   "source": [
    "all_keys = list(clade_stats_df.keys())[2:]\n",
    "\n",
    "X = clade_stats_df[all_keys]\n",
    "y = clade_stats_df['logistic_growth']\n",
    "\n",
    "lin_reg = linear_model.LinearRegression()\n",
    "\n",
    "# number of features to select\n",
    "sel_num = 20\n",
    "selected_feature = []\n",
    "\n",
    "# Split into 5 training and test sets\n",
    "rs = ShuffleSplit(n_splits=5, test_size=.25, random_state=0)\n",
    "\n",
    "for i in all_keys:\n",
    "\n",
    "    for train_index, test_index in rs.split(X):\n",
    "        X_train, X_valid = X[i][train_index], X[i][test_index]\n",
    "        y_train, y_valid = y[train_index], y[test_index]\n",
    "\n",
    "        \n",
    "        model_train = lin_reg.fit(np.array(X_train).reshape(-1, 1), np.array(y_train).reshape(-1, 1))\n",
    "        predictions_train = model_train.predict(np.array(X_valid).reshape(-1, 1))\n",
    "        print(mean_squared_error(np.array(y_valid).reshape(-1, 1), predictions_train))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clade</th>\n",
       "      <th>logistic_growth</th>\n",
       "      <th>num_s1_muts</th>\n",
       "      <th>N_204G</th>\n",
       "      <th>N_203R</th>\n",
       "      <th>N_203K</th>\n",
       "      <th>S_681P</th>\n",
       "      <th>S_681H</th>\n",
       "      <th>S_681R</th>\n",
       "      <th>S_501N</th>\n",
       "      <th>...</th>\n",
       "      <th>S_570V</th>\n",
       "      <th>ORF1a_2230T</th>\n",
       "      <th>S_484G</th>\n",
       "      <th>N_205-</th>\n",
       "      <th>S_417T</th>\n",
       "      <th>N_199-</th>\n",
       "      <th>N_199Q</th>\n",
       "      <th>ORF1b_2613S</th>\n",
       "      <th>S_701T</th>\n",
       "      <th>S_452Q</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1052</th>\n",
       "      <td>NODE_0000055</td>\n",
       "      <td>4.095786</td>\n",
       "      <td>1</td>\n",
       "      <td>0.007332</td>\n",
       "      <td>0.006180</td>\n",
       "      <td>0.992135</td>\n",
       "      <td>0.565900</td>\n",
       "      <td>0.424565</td>\n",
       "      <td>0.008974</td>\n",
       "      <td>0.598519</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000552</td>\n",
       "      <td>0.361738</td>\n",
       "      <td>0.00057</td>\n",
       "      <td>0.00056</td>\n",
       "      <td>0.023889</td>\n",
       "      <td>0.00056</td>\n",
       "      <td>0.00056</td>\n",
       "      <td>0.001147</td>\n",
       "      <td>0.000562</td>\n",
       "      <td>0.011824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1053</th>\n",
       "      <td>NODE_0000056</td>\n",
       "      <td>4.095786</td>\n",
       "      <td>1</td>\n",
       "      <td>0.006772</td>\n",
       "      <td>0.006183</td>\n",
       "      <td>0.992130</td>\n",
       "      <td>0.565657</td>\n",
       "      <td>0.424804</td>\n",
       "      <td>0.008979</td>\n",
       "      <td>0.598291</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000553</td>\n",
       "      <td>0.361942</td>\n",
       "      <td>0.00057</td>\n",
       "      <td>0.00056</td>\n",
       "      <td>0.023902</td>\n",
       "      <td>0.00056</td>\n",
       "      <td>0.00056</td>\n",
       "      <td>0.001148</td>\n",
       "      <td>0.000562</td>\n",
       "      <td>0.011831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1054</th>\n",
       "      <td>NODE_0001698</td>\n",
       "      <td>4.095786</td>\n",
       "      <td>2</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.923077</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055</th>\n",
       "      <td>NODE_0002319</td>\n",
       "      <td>4.095786</td>\n",
       "      <td>2</td>\n",
       "      <td>0.090909</td>\n",
       "      <td>0.090909</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1056</th>\n",
       "      <td>NODE_0002322</td>\n",
       "      <td>4.095786</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2083</th>\n",
       "      <td>NODE_0004144</td>\n",
       "      <td>4.111753</td>\n",
       "      <td>6</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2084</th>\n",
       "      <td>NODE_0004145</td>\n",
       "      <td>4.111753</td>\n",
       "      <td>6</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2085</th>\n",
       "      <td>NODE_0000645</td>\n",
       "      <td>4.111753</td>\n",
       "      <td>6</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2086</th>\n",
       "      <td>NODE_0004147</td>\n",
       "      <td>4.111753</td>\n",
       "      <td>6</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2087</th>\n",
       "      <td>NODE_0004148</td>\n",
       "      <td>4.111753</td>\n",
       "      <td>6</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>489 rows Ã— 178 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             clade  logistic_growth  num_s1_muts    N_204G    N_203R  \\\n",
       "1052  NODE_0000055         4.095786            1  0.007332  0.006180   \n",
       "1053  NODE_0000056         4.095786            1  0.006772  0.006183   \n",
       "1054  NODE_0001698         4.095786            2  0.076923  0.076923   \n",
       "1055  NODE_0002319         4.095786            2  0.090909  0.090909   \n",
       "1056  NODE_0002322         4.095786            2  0.000000  0.000000   \n",
       "...            ...              ...          ...       ...       ...   \n",
       "2083  NODE_0004144         4.111753            6  0.000000  0.000000   \n",
       "2084  NODE_0004145         4.111753            6  0.000000  0.000000   \n",
       "2085  NODE_0000645         4.111753            6  0.000000  0.000000   \n",
       "2086  NODE_0004147         4.111753            6  0.000000  0.000000   \n",
       "2087  NODE_0004148         4.111753            6  0.000000  0.000000   \n",
       "\n",
       "        N_203K    S_681P    S_681H    S_681R    S_501N  ...    S_570V  \\\n",
       "1052  0.992135  0.565900  0.424565  0.008974  0.598519  ...  0.000552   \n",
       "1053  0.992130  0.565657  0.424804  0.008979  0.598291  ...  0.000553   \n",
       "1054  0.923077  1.000000  0.000000  0.000000  1.000000  ...  0.000000   \n",
       "1055  0.909091  1.000000  0.000000  0.000000  1.000000  ...  0.000000   \n",
       "1056  1.000000  1.000000  0.000000  0.000000  1.000000  ...  0.000000   \n",
       "...        ...       ...       ...       ...       ...  ...       ...   \n",
       "2083  1.000000  0.000000  1.000000  0.000000  0.000000  ...  0.000000   \n",
       "2084  1.000000  0.000000  1.000000  0.000000  0.000000  ...  0.000000   \n",
       "2085  1.000000  0.000000  1.000000  0.000000  0.000000  ...  0.000000   \n",
       "2086  1.000000  0.000000  1.000000  0.000000  0.000000  ...  0.000000   \n",
       "2087  1.000000  0.000000  1.000000  0.000000  0.000000  ...  0.000000   \n",
       "\n",
       "      ORF1a_2230T   S_484G   N_205-    S_417T   N_199-   N_199Q  ORF1b_2613S  \\\n",
       "1052     0.361738  0.00057  0.00056  0.023889  0.00056  0.00056     0.001147   \n",
       "1053     0.361942  0.00057  0.00056  0.023902  0.00056  0.00056     0.001148   \n",
       "1054     0.000000  0.00000  0.00000  0.000000  0.00000  0.00000     0.000000   \n",
       "1055     0.000000  0.00000  0.00000  0.000000  0.00000  0.00000     0.000000   \n",
       "1056     0.000000  0.00000  0.00000  0.000000  0.00000  0.00000     0.000000   \n",
       "...           ...      ...      ...       ...      ...      ...          ...   \n",
       "2083     1.000000  0.00000  0.00000  0.000000  0.00000  0.00000     0.000000   \n",
       "2084     1.000000  0.00000  0.00000  0.000000  0.00000  0.00000     0.000000   \n",
       "2085     1.000000  0.00000  0.00000  0.000000  0.00000  0.00000     0.000000   \n",
       "2086     1.000000  0.00000  0.00000  0.000000  0.00000  0.00000     0.000000   \n",
       "2087     1.000000  0.00000  0.00000  0.000000  0.00000  0.00000     0.000000   \n",
       "\n",
       "        S_701T    S_452Q  \n",
       "1052  0.000562  0.011824  \n",
       "1053  0.000562  0.011831  \n",
       "1054  0.000000  0.000000  \n",
       "1055  0.000000  0.000000  \n",
       "1056  0.000000  0.000000  \n",
       "...        ...       ...  \n",
       "2083  0.000000  0.000000  \n",
       "2084  0.000000  0.000000  \n",
       "2085  0.000000  0.000000  \n",
       "2086  0.000000  0.000000  \n",
       "2087  0.000000  0.000000  \n",
       "\n",
       "[489 rows x 178 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clade_stats_df[clade_stats_df['logistic_growth']>=4.0]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "105     0.134182\n",
      "106     0.846154\n",
      "107     0.916667\n",
      "109     1.000000\n",
      "110     1.000000\n",
      "          ...   \n",
      "2083    1.000000\n",
      "2084    1.000000\n",
      "2085    1.000000\n",
      "2086    1.000000\n",
      "2087    1.000000\n",
      "Name: ORF1a_3675-, Length: 782, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "#add in nsp6 deletion! this is ORF1a\t3675\n",
    "print(clade_stats_df[clade_stats_df['ORF1a_3675-'].notnull()]['ORF1a_3675-'])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#node.__dict__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, make dataframe containing information about number of S1 mutations from root to clade and all mutations that occurred between root and clade. Each row is a clade"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#initiate list to store all clade history info\n",
    "clades_history = []\n",
    "\n",
    "#Function to find path from root to clade\n",
    "def get_parent(tree, child_clade):\n",
    "    node_path = tree.get_path(child_clade)\n",
    "    return node_path\n",
    "\n",
    "\n",
    "#Only want to look at clades, don't care about tips\n",
    "for node in tree.find_clades(terminal=False):\n",
    "    \n",
    "    #find all mutations that occurred on path from root to clade\n",
    "    root_to_clade_mutations = []\n",
    "    for parent in get_parent(tree, node):\n",
    "        if hasattr(parent, 'branch_attrs'):\n",
    "            root_to_clade_mutations.append(parent.branch_attrs['mutations'])\n",
    "    \n",
    "    #flatten root_to_clade_mutations, making dict entry for nucleotide muts and amino acid subs in each gene\n",
    "    possible_mutation_site = ['ORF1a', 'ORF1b', 'S', 'ORF3a', 'E', 'M', 'ORF6', \n",
    "                              'ORF7a', 'ORF7b', 'ORF8', 'ORF9b', 'N', 'nuc']\n",
    "    mutations_on_path = {k:[] for k in possible_mutation_site}\n",
    "    \n",
    "    for parent_muts in root_to_clade_mutations:\n",
    "        for k,v in parent_muts.items():\n",
    "            mutations_on_path[k]+=v\n",
    "    \n",
    "    \n",
    "    #all S1 muts from root to clade already stored as json value\n",
    "    if \"S1_mutations\" in node.node_attrs:\n",
    "        s1_mutations = node.node_attrs[\"S1_mutations\"][\"value\"]\n",
    "    else:\n",
    "        s1_mutations = None\n",
    "    \n",
    "    #find stored logistic growth value\n",
    "    if \"logistic_growth\" in node.node_attrs:\n",
    "        logistic_growth = node.node_attrs[\"logistic_growth\"][\"value\"]\n",
    "    else:\n",
    "        logistic_growth = None\n",
    "    \n",
    "    #copy dictionary and add key/values for clade name and number of s1 mutations and logistic growth\n",
    "    clade_mutation_history = mutations_on_path\n",
    "    clade_mutation_history['clade'] = node.name\n",
    "    clade_mutation_history['num_s1_mutations'] = s1_mutations\n",
    "    clade_mutation_history['logistic_growth'] = logistic_growth\n",
    "    \n",
    "\n",
    "    \n",
    "    clades_history.append(clade_mutation_history)\n",
    "    \n",
    "\n",
    "\n",
    "#turn list of clade history info into a dataframe\n",
    "clades_df = pd.DataFrame(clades_history)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ORF1a</th>\n",
       "      <th>ORF1b</th>\n",
       "      <th>S</th>\n",
       "      <th>ORF3a</th>\n",
       "      <th>E</th>\n",
       "      <th>M</th>\n",
       "      <th>ORF6</th>\n",
       "      <th>ORF7a</th>\n",
       "      <th>ORF7b</th>\n",
       "      <th>ORF8</th>\n",
       "      <th>ORF9b</th>\n",
       "      <th>N</th>\n",
       "      <th>nuc</th>\n",
       "      <th>clade</th>\n",
       "      <th>num_s1_mutations</th>\n",
       "      <th>logistic_growth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>292</th>\n",
       "      <td>[S3675-, G3676-, F3677-]</td>\n",
       "      <td>[P314L]</td>\n",
       "      <td>[D614G, E484K]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[I82T]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[T205I]</td>\n",
       "      <td>[C3037T, C14408T, C241T, A23403G, A21993-, T21...</td>\n",
       "      <td>NODE_0001779</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6.511755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293</th>\n",
       "      <td>[S3675-, G3676-, F3677-, A2123V, E2607K, M3752I]</td>\n",
       "      <td>[P314L]</td>\n",
       "      <td>[D614G, E484K, I210T, D936N, S939F, T1027I]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[I82T]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[E22D]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[P10S]</td>\n",
       "      <td>[T205I, P13L, S201I]</td>\n",
       "      <td>[C3037T, C14408T, C241T, A23403G, A21993-, T21...</td>\n",
       "      <td>NODE_0001780</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6.511755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294</th>\n",
       "      <td>[S3675-, G3676-, F3677-, A2123V, E2607K, M3752I]</td>\n",
       "      <td>[P314L]</td>\n",
       "      <td>[D614G, E484K, I210T, D936N, S939F, T1027I, N4...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[I82T]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[E22D]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[P10S]</td>\n",
       "      <td>[T205I, P13L, S201I]</td>\n",
       "      <td>[C3037T, C14408T, C241T, A23403G, A21993-, T21...</td>\n",
       "      <td>NODE_0001781</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.511755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>[S3675-, G3676-, F3677-, T2007I]</td>\n",
       "      <td>[P314L]</td>\n",
       "      <td>[D614G, E484K, A67V, H69-, V70-, Y144-, Q677H]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[L21F]</td>\n",
       "      <td>[I82T]</td>\n",
       "      <td>[F2-]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[H9D]</td>\n",
       "      <td>[T205I, S2M, D3Y, A12G]</td>\n",
       "      <td>[C3037T, C14408T, C241T, A23403G, A21993-, T21...</td>\n",
       "      <td>NODE_0001782</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.511755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>[S3675-, G3676-, F3677-, T2007I]</td>\n",
       "      <td>[P314L, L314F]</td>\n",
       "      <td>[D614G, E484K, A67V, H69-, V70-, Y144-, Q677H,...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[L21F]</td>\n",
       "      <td>[I82T]</td>\n",
       "      <td>[F2-]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[H9D]</td>\n",
       "      <td>[T205I, S2M, D3Y, A12G]</td>\n",
       "      <td>[C3037T, C14408T, C241T, A23403G, A21993-, T21...</td>\n",
       "      <td>NODE_0001783</td>\n",
       "      <td>7.0</td>\n",
       "      <td>6.511755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3275</th>\n",
       "      <td>[I2230T, S3675-, G3676-, F3677-, T1001I, A1708D]</td>\n",
       "      <td>[P314L, K1383R, P1001S]</td>\n",
       "      <td>[D614G, N501Y, H69-, V70-, Y144-, S982A, A570D...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Q27*, R52I, Y73C, K68*, *68K]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[R203K, G204R, S235F, D3L]</td>\n",
       "      <td>[C3037T, C14408T, C241T, A23403G, G28881A, G28...</td>\n",
       "      <td>NODE_0004177</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.519968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3276</th>\n",
       "      <td>[I2230T, S3675-, G3676-, F3677-, T1001I, A1708D]</td>\n",
       "      <td>[P314L, K1383R, P1001S]</td>\n",
       "      <td>[D614G, N501Y, H69-, V70-, Y144-, S982A, A570D...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Q27*, R52I, Y73C, K68*, *68K]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[R203K, G204R, S235F, D3L]</td>\n",
       "      <td>[C3037T, C14408T, C241T, A23403G, G28881A, G28...</td>\n",
       "      <td>NODE_0004178</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.519968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3277</th>\n",
       "      <td>[I2230T, S3675-, G3676-, F3677-, T1001I, A1708D]</td>\n",
       "      <td>[P314L, K1383R, P1001S]</td>\n",
       "      <td>[D614G, N501Y, H69-, V70-, Y144-, S982A, A570D...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Q27*, R52I, Y73C, K68*, *68K]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[R203K, G204R, S235F, D3L]</td>\n",
       "      <td>[C3037T, C14408T, C241T, A23403G, G28881A, G28...</td>\n",
       "      <td>NODE_0004179</td>\n",
       "      <td>6.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3278</th>\n",
       "      <td>[I2230T, S3675-, G3676-, F3677-, T1001I, A1708D]</td>\n",
       "      <td>[P314L, K1383R, P1001S, I1181S]</td>\n",
       "      <td>[D614G, N501Y, H69-, V70-, Y144-, S982A, A570D...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Q27*, R52I, Y73C, K68*, *68K]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[R203K, G204R, S235F, D3L]</td>\n",
       "      <td>[C3037T, C14408T, C241T, A23403G, G28881A, G28...</td>\n",
       "      <td>NODE_0000816</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.519968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3279</th>\n",
       "      <td>[I2230T, S3675-, G3676-, F3677-, T1001I, A1708D]</td>\n",
       "      <td>[P314L, K1383R, P1001S, I1181S]</td>\n",
       "      <td>[D614G, N501Y, H69-, V70-, Y144-, S982A, A570D...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Q27*, R52I, Y73C, K68*, *68K]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[R203K, G204R, S235F, D3L]</td>\n",
       "      <td>[C3037T, C14408T, C241T, A23403G, G28881A, G28...</td>\n",
       "      <td>NODE_0000817</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.519968</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>753 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 ORF1a  \\\n",
       "292                           [S3675-, G3676-, F3677-]   \n",
       "293   [S3675-, G3676-, F3677-, A2123V, E2607K, M3752I]   \n",
       "294   [S3675-, G3676-, F3677-, A2123V, E2607K, M3752I]   \n",
       "295                   [S3675-, G3676-, F3677-, T2007I]   \n",
       "296                   [S3675-, G3676-, F3677-, T2007I]   \n",
       "...                                                ...   \n",
       "3275  [I2230T, S3675-, G3676-, F3677-, T1001I, A1708D]   \n",
       "3276  [I2230T, S3675-, G3676-, F3677-, T1001I, A1708D]   \n",
       "3277  [I2230T, S3675-, G3676-, F3677-, T1001I, A1708D]   \n",
       "3278  [I2230T, S3675-, G3676-, F3677-, T1001I, A1708D]   \n",
       "3279  [I2230T, S3675-, G3676-, F3677-, T1001I, A1708D]   \n",
       "\n",
       "                                ORF1b  \\\n",
       "292                           [P314L]   \n",
       "293                           [P314L]   \n",
       "294                           [P314L]   \n",
       "295                           [P314L]   \n",
       "296                    [P314L, L314F]   \n",
       "...                               ...   \n",
       "3275          [P314L, K1383R, P1001S]   \n",
       "3276          [P314L, K1383R, P1001S]   \n",
       "3277          [P314L, K1383R, P1001S]   \n",
       "3278  [P314L, K1383R, P1001S, I1181S]   \n",
       "3279  [P314L, K1383R, P1001S, I1181S]   \n",
       "\n",
       "                                                      S ORF3a       E       M  \\\n",
       "292                                      [D614G, E484K]    []      []  [I82T]   \n",
       "293         [D614G, E484K, I210T, D936N, S939F, T1027I]    []      []  [I82T]   \n",
       "294   [D614G, E484K, I210T, D936N, S939F, T1027I, N4...    []      []  [I82T]   \n",
       "295      [D614G, E484K, A67V, H69-, V70-, Y144-, Q677H]    []  [L21F]  [I82T]   \n",
       "296   [D614G, E484K, A67V, H69-, V70-, Y144-, Q677H,...    []  [L21F]  [I82T]   \n",
       "...                                                 ...   ...     ...     ...   \n",
       "3275  [D614G, N501Y, H69-, V70-, Y144-, S982A, A570D...    []      []      []   \n",
       "3276  [D614G, N501Y, H69-, V70-, Y144-, S982A, A570D...    []      []      []   \n",
       "3277  [D614G, N501Y, H69-, V70-, Y144-, S982A, A570D...    []      []      []   \n",
       "3278  [D614G, N501Y, H69-, V70-, Y144-, S982A, A570D...    []      []      []   \n",
       "3279  [D614G, N501Y, H69-, V70-, Y144-, S982A, A570D...    []      []      []   \n",
       "\n",
       "       ORF6   ORF7a ORF7b                            ORF8   ORF9b  \\\n",
       "292      []      []    []                              []      []   \n",
       "293      []  [E22D]    []                              []  [P10S]   \n",
       "294      []  [E22D]    []                              []  [P10S]   \n",
       "295   [F2-]      []    []                              []   [H9D]   \n",
       "296   [F2-]      []    []                              []   [H9D]   \n",
       "...     ...     ...   ...                             ...     ...   \n",
       "3275     []      []    []  [Q27*, R52I, Y73C, K68*, *68K]      []   \n",
       "3276     []      []    []  [Q27*, R52I, Y73C, K68*, *68K]      []   \n",
       "3277     []      []    []  [Q27*, R52I, Y73C, K68*, *68K]      []   \n",
       "3278     []      []    []  [Q27*, R52I, Y73C, K68*, *68K]      []   \n",
       "3279     []      []    []  [Q27*, R52I, Y73C, K68*, *68K]      []   \n",
       "\n",
       "                               N  \\\n",
       "292                      [T205I]   \n",
       "293         [T205I, P13L, S201I]   \n",
       "294         [T205I, P13L, S201I]   \n",
       "295      [T205I, S2M, D3Y, A12G]   \n",
       "296      [T205I, S2M, D3Y, A12G]   \n",
       "...                          ...   \n",
       "3275  [R203K, G204R, S235F, D3L]   \n",
       "3276  [R203K, G204R, S235F, D3L]   \n",
       "3277  [R203K, G204R, S235F, D3L]   \n",
       "3278  [R203K, G204R, S235F, D3L]   \n",
       "3279  [R203K, G204R, S235F, D3L]   \n",
       "\n",
       "                                                    nuc         clade  \\\n",
       "292   [C3037T, C14408T, C241T, A23403G, A21993-, T21...  NODE_0001779   \n",
       "293   [C3037T, C14408T, C241T, A23403G, A21993-, T21...  NODE_0001780   \n",
       "294   [C3037T, C14408T, C241T, A23403G, A21993-, T21...  NODE_0001781   \n",
       "295   [C3037T, C14408T, C241T, A23403G, A21993-, T21...  NODE_0001782   \n",
       "296   [C3037T, C14408T, C241T, A23403G, A21993-, T21...  NODE_0001783   \n",
       "...                                                 ...           ...   \n",
       "3275  [C3037T, C14408T, C241T, A23403G, G28881A, G28...  NODE_0004177   \n",
       "3276  [C3037T, C14408T, C241T, A23403G, G28881A, G28...  NODE_0004178   \n",
       "3277  [C3037T, C14408T, C241T, A23403G, G28881A, G28...  NODE_0004179   \n",
       "3278  [C3037T, C14408T, C241T, A23403G, G28881A, G28...  NODE_0000816   \n",
       "3279  [C3037T, C14408T, C241T, A23403G, G28881A, G28...  NODE_0000817   \n",
       "\n",
       "      num_s1_mutations  logistic_growth  \n",
       "292                2.0         6.511755  \n",
       "293                3.0         6.511755  \n",
       "294                4.0         6.511755  \n",
       "295                6.0         6.511755  \n",
       "296                7.0         6.511755  \n",
       "...                ...              ...  \n",
       "3275               6.0         5.519968  \n",
       "3276               6.0         5.519968  \n",
       "3277               6.0              NaN  \n",
       "3278               6.0         5.519968  \n",
       "3279               6.0         5.519968  \n",
       "\n",
       "[753 rows x 16 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clades_df[pd.DataFrame(clades_df.nuc.tolist()).isin(['T11288-']).any(1).values]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ncov-kernel",
   "language": "python",
   "name": "ncov-kernel"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
